core/src/HIP/Kokkos_HIP_Team.hpp

```diff
@@ -774,16 +774,15 @@ KOKKOS_INLINE_FUNCTION void parallel_scan(
       FunctorType>::value_type;
 
   const auto start     = loop_bounds.start;
-  const auto end       = loop_bounds.end + loop_bounds.start;
+  const auto end       = loop_bounds.end;
   auto& member         = loop_bounds.member;
   const auto team_size = member.team_size();
   const auto team_rank = member.team_rank();
-  const auto nchunk    = (end - start) / loop_bounds.member.team_size();
+  const auto nchunk    = (end - start + team_size - 1) / team_size;
   value_type accum     = 0;
-
   // each team has to process one or more chunks of the prefix scan
-  for (iType i = 0; i < nchunk + 1; ++i) {
-    auto ii = i * team_size + team_rank;
+  for (iType i = 0; i < nchunk; ++i) {
+    auto ii = start + i * team_size + team_rank;
```

Comment:
Maybe a comment somewhere indicating that this is exactly the same as the Cuda version would be helpful, so that it's clear these changes are supposed to be exactly the same in both places? I guess it's the next best thing to code reuse, which would be far more maintainable than this...

Reply:
The OpenMPTarget stuff will not long term be the same code, in fact the parallel for and reduce already work differently than the CUDA backend since they use direct pragmas instead of our own loop splitting algorithm. For scan that will also come as soon as compilers implement it. The HIP stuff is the same, but that is true for significant chunks of the HIP backend. I don't feel like figuring out how to best share code between the two right now. Also I am not sure that that is totally in our long term interest. The shared code between the OpenMP CPU and the Pthreads backend for example made it harder for us to use native OpenMP constructs in the CPU backend, and in order to do so we now have to untangle the backends again ...

Reply:
added comments

Reply:
We can continue to litigate this elsewhere; I won't hold up this PR based on this, but I want to register my disagreement here.

> The OpenMPTarget stuff will not long term be the same code, in fact the parallel for and reduce already work differently

That's why it's absolutely essential to use shared code where things are the same—so that we can quickly pick out what parts of the OpenMPTarget will also change and what parts won't when we come across a bug like this.

> Also I am not sure that that is totally in our long term interest. 

I strongly disagree with this, and I don't know of any general software engineering advice that supports this position. 

> The shared code between the OpenMP CPU and the Pthreads backend for example made it harder for us to use native OpenMP constructs in the CPU backend,

This is because a lot of that sharing was done poorly, and things there are way more fancy than necessary. I'm not asking for that. I'm just asking for a simple function template or CRTP base class. Something to express that "these pieces of code are the exact same tokens in two different places".

> and in order to do so we now have to untangle the backends again ...

No, especially if this is done right (like, for instance, in the default graph implementation), we just add an extra specialization for new stuff.